### PR TITLE
Cell Maintenance: Spotbugs & TCP

### PR DESCRIPTION
**Pull request by schwepmo**
## Type of this PR 

- [x] Bug fix
- [x] Enhancement

## Description

What is this PR about?

* This PR resolves some issues in the cell module, that couldn't be addressed before release
* `CellMessageRoutingBuilder` now has the method `protocol(...)` to specify the used protocol for a message, by default UDP will be used
* an integration test formerly contained in mosaic-extended is now moved to mosaic (SendAndReceiveIT) and extended by some tcp acknowledgement tests
* some Spotbugs Errors in configurations are now properly suppressed
## Issue(s) related to this PR

 * Resolves internal issues 44, 52, 57
 
## Affected parts of the online documentation

* the new `protocol(...)`-method in the `CellMessageRoutingBuilder` is documented

 
## Definition of Done

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] All commits are signed-off (`-s`) with your mail address of your Eclipse Account.
- [x] `origin/main` has been merged into your Fork.
- [x] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [x] There are no new SpotBugs warnings. 
- [x] Coding guidelines have been observed (see CONTRIBUTING.md).
- [x] All tests pass.

### CODE REVIEW

File: fed/mosaic-cell/src/main/java/org/eclipse/mosaic/fed/cell/utility/CapacityUtility.java [hunk 1]
```diff
@@ -31,6 +32,7 @@
 /**
  * Static methods for handling the bandwidth calculation.
  */
+@SuppressWarnings(value = {"NP_UNWRITTEN_PUBLIC_OR_PROTECTED_FIELD"}, justification = "filled by json")
```
Comment (kschrab) [inline fed/mosaic-cell/src/main/java/org/eclipse/mosaic/fed/cell/utility/CapacityUtility.java:35]: Wenn mans genau nimmt, ist "GSON" derjenige der das befüllt :)

File: lib/mosaic-objects/src/main/java/org/eclipse/mosaic/lib/objects/addressing/CellMessageRoutingBuilder.java [hunk 2]
```diff
@@ -131,5 +143,15 @@ public MessageRouting topoCast(String name) {
...
+    /**
+     * Sets the {@link ProtocolType} for the routing.
+     *
+     * @param protocolType the {@link ProtocolType} to be used
+     * @return the {@link CellMessageRoutingBuilder}
+     */
+    public CellMessageRoutingBuilder protocol(ProtocolType protocolType) {
+        this.protocolType = protocolType;
+        return this;
+    }
```
Comment (kschrab) [inline lib/mosaic-objects/src/main/java/org/eclipse/mosaic/lib/objects/addressing/CellMessageRoutingBuilder.java:155]: Hier hätte ich es sogar noch schön gefunden, wenn es noch die Methoden `tcp()` und `udp()` gibt:

```java
public CellMessageRoutingBuilder tcp() {
   return protocol(ProtocolType.TCP);
}

public CellMessageRoutingBuilder udp() {
   return protocol(ProtocolType.UDP);
}
```

File: fed/mosaic-cell/src/main/java/org/eclipse/mosaic/fed/cell/utility/CapacityUtility.java (hunk 1 above)
Comment (schwepmo) [reply fed/mosaic-cell/src/main/java/org/eclipse/mosaic/fed/cell/utility/CapacityUtility.java]: fixed.

File: lib/mosaic-objects/src/main/java/org/eclipse/mosaic/lib/objects/addressing/CellMessageRoutingBuilder.java (hunk 2 above)
Comment (schwepmo) [reply lib/mosaic-objects/src/main/java/org/eclipse/mosaic/lib/objects/addressing/CellMessageRoutingBuilder.java:155]: Added.

File: test/mosaic-integration-tests/pom.xml [hunk 3]
```diff
@@ -53,6 +53,19 @@
             <version>${mosaic.version}</version>
             <scope>test</scope>
         </dependency>
+        <!--For applications-->
```
Comment (schwepmo) [inline test/mosaic-integration-tests/pom.xml:56]: @kschrab dependencies fine like this?

File: test/mosaic-integration-tests/src/test/java/org/eclipse/mosaic/test/CommunicationSimulatorsIT.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+ *     <li/> TCP acknowledgement test
+ * </ul>
+ * Units have the following roles:
+ * <ul>
+ *     <li/> veh0: Sending Cams
+ *     <li/> veh1: Receiving and returning RoundTripMessage
+ *     <li/> veh2: Sending Cams
+ * </ul>
+ */
+public class CommunicationSimulatorsIT {
```
Comment (schwepmo) [inline test/mosaic-integration-tests/src/test/java/org/eclipse/mosaic/test/CommunicationSimulatorsIT.java:49]: @kschrab Naming okay?

File: test/mosaic-integration-tests/pom.xml (hunk 3 above)
Comment (schwepmo) [reply test/mosaic-integration-tests/pom.xml:56]: Apparently not since build fails 😅

Comment (schwepmo) [reply test/mosaic-integration-tests/pom.xml:56]: At the moment I just put the applications for the new integration test into the package `org.eclipse.mosaic.test.sendandreceive`, which requires the new dependencies. Should the apps rather be put in a separate module, which is then added as a dependency?

Comment (schwepmo) [reply test/mosaic-integration-tests/pom.xml:56]: Alright. Seems to be fixed now.

Comment (kschrab) [reply test/mosaic-integration-tests/pom.xml:56]: `org.eclipse.mosaic.test.app.sendandreceive`

und vllt den test lieber "SendAndReceiveIT" oder vllt "CellSnsSendAndReceiveIT" nennen."CommunicationSimulatorsIT" finde ich etwas zu allgemein, OMNeT++/ns-3 sind ja nicht dabei.

File: test/scenarios/CommunicationSimulators/cell/cell_config.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{
+    "networkConfigurationFile": "network.json",
+	"regionConfigurationFile":  "regions.json"
```
Comment (kschrab) [inline test/scenarios/CommunicationSimulators/cell/cell_config.json:3]: tabs -> spaces

File: test/scenarios/CommunicationSimulators/output/output_config.xml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        <synchronized>true</synchronized>
+        <host>local</host>
+        <port>50500</port>
+        <subscriptions>
+        <subscription id="TrafficLightRegistration"></subscription>
+        <subscription id="RsuRegistration"></subscription>
+        <subscription id="VehicleRegistration"></subscription>
+        <subscription id="VehicleUpdates"></subscription>
+        </subscriptions>
+    </output>
```
Comment (kschrab) [inline test/scenarios/CommunicationSimulators/output/output_config.xml:32]: remove

File: test/scenarios/CommunicationSimulators/output/output_config.xml [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                    <entry convertvalue="vsimrti_time_converter">Updated:Time</entry>
+                    <entry>"1004000000100000001"</entry>
+                    <entry convertvalue="1000000">Updated:Position.Latitude</entry>
+                    <entry convertvalue="1000000">Updated:Position.Longitude</entry>
+                    <entry convertvalue="10000">Updated:Direction</entry>
+                    <entry convertvalue="1000">Updated:Speed</entry>
+                </entries>
+            </subscription>
+        </subscriptions>
+    </output>
```
Comment (kschrab) [inline test/scenarios/CommunicationSimulators/output/output_config.xml:68]: remove

File: test/scenarios/CommunicationSimulators/sumo/tiergarten.rou.xml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" ?>
+<!-- Generated by VSimRTI -->
```
Comment (kschrab) [inline test/scenarios/CommunicationSimulators/sumo/tiergarten.rou.xml:2]: remove "generated by VSimRTI"

tabs -> spaces in all xml-files (nod, edg, con, rou.xml)

File: test/mosaic-integration-tests/pom.xml [hunk 9]
```diff
@@ -53,6 +53,23 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.eclipse.mosaic</groupId>
+            <artifactId>mosaic-objects</artifactId>
+            <version>${mosaic.version}</version>
+        </dependency>
+
     </dependencies>
 
     <build>
```
Comment (kschrab) [inline test/mosaic-integration-tests/pom.xml:75]: Our mosaic-extended project requires the Send/Receive applications as well for some integration tests. Those are currently failing as you moved them to eclipse-mosaic. We need to make them available for those integration tests as well. This is how you can achieve this:

1) Move all apps into the package `org.eclipse.mosaic.test.app....` (you still can use additionals subpackages)
2) add an `<include>org/eclipse/mosaic/test/app/**</include>` to the `create-test-jar-with-rules` section below
3) For the legacy integration tests in mosaic-extended, you have to copy the `target/mosaic-integration-tests-20.1-SNAPSHOT-tests.jar` to `mosaic/lib/mosaic` by adjusting the `setup_testing.sh`.

File: test/mosaic-integration-tests/src/test/java/org/eclipse/mosaic/test/CommunicationSimulatorsIT.java (hunk 4 above)
Comment (kschrab) [reply test/mosaic-integration-tests/src/test/java/org/eclipse/mosaic/test/CommunicationSimulatorsIT.java:49]: I would prefer "SendAndReceiveIT" or even "CellSnsSendAndReceiveIT", as "CommunicationSimulatorsIT" is too general ( e.g. OMNeT++/ns-3 are not tested here).

**Review by kschrab — CHANGES_REQUESTED** [review]
See comments in code

File: test/mosaic-integration-tests/src/test/java/org/eclipse/mosaic/test/CommunicationSimulatorsIT.java (hunk 4 above)
Comment (schwepmo) [reply test/mosaic-integration-tests/src/test/java/org/eclipse/mosaic/test/CommunicationSimulatorsIT.java:49]: Well there is some communication via the SNS as well, so I would prefer `SendAndReceiveIT`. Should the scenario be called
``SendAndReceive` analogously?

File: test/mosaic-integration-tests/pom.xml (hunk 9 above)
Comment (schwepmo) [reply test/mosaic-integration-tests/pom.xml:75]: Am I right in assuming, that the `mapping_config.json`'s have to be adjusted to `org.eclipse.mosaic.test.app...`

Comment (kschrab) [reply test/mosaic-integration-tests/pom.xml:75]: Correct

File: test/mosaic-integration-tests/src/test/java/org/eclipse/mosaic/test/CommunicationSimulatorsIT.java (hunk 4 above)
Comment (kschrab) [reply test/mosaic-integration-tests/src/test/java/org/eclipse/mosaic/test/CommunicationSimulatorsIT.java]: Yes

File: test/mosaic-integration-tests/src/test/java/org/eclipse/mosaic/test/SendAndReceiveIT.java [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+        LogAssert.contains(
+                simulationRule,
+                VEH_ROUND_TRIP,
+                ".*Received acknowledgement for round trip message #[0-9]+ and \\[acknowledged=true\\] \\(at simulation time "
+                        + TIME.format(timeOfSending + timeFromTmcToVeh + timeFromVehToTmc) + "\\).*"
+        );
+    }
+
+    @Test
+    public void tcpNacksAreSent() throws Exception {
```
Comment (schwepmo) [inline test/mosaic-integration-tests/src/test/java/org/eclipse/mosaic/test/SendAndReceiveIT.java:240]: Right now I just added two basic nack-assertions, could be extendend

**Review by kschrab — APPROVED** [review]

**Pull request merged**